pkg/webhook/webhook.go

```diff
@@ -60,74 +59,127 @@ type WebHook struct {
+
+func init() {
+	flag.StringVar(&userConfig.webhookConfigName, "webhook-config-name", "spark-webhook-config", "The name of the MutatingWebhookConfiguration object to create.")
+	flag.StringVar(&userConfig.serverCert, "webhook-server-cert", "/etc/webhook-certs/server-cert.pem", "Path to the X.509-formatted webhook certificate.")
+	flag.StringVar(&userConfig.serverCertKey, "webhook-server-cert-key", "/etc/webhook-certs/server-key.pem", "Path to the webhook certificate key.")
+	flag.StringVar(&userConfig.caCert, "webhook-ca-cert", "/etc/webhook-certs/ca-cert.pem", "Path to the X.509-formatted webhook CA certificate.")
+	flag.DurationVar(&userConfig.certReloadInterval, "webhook-cert-reload-interval", 15*time.Minute, "Time between webhook cert reloads.")
+	flag.StringVar(&userConfig.webhookServiceNamespace, "webhook-svc-namespace", "spark-operator", "The namespace of the Service for the webhook server.")
+	flag.StringVar(&userConfig.webhookServiceName, "webhook-svc-name", "spark-webhook", "The name of the Service for the webhook server.")
+	flag.IntVar(&userConfig.webhookPort, "webhook-port", 8080, "Service port of the webhook server.")
+	flag.BoolVar(&userConfig.webhookFailOnError, "webhook-fail-on-error", false, "Whether Kubernetes should reject requests when the webhook fails.")
+	flag.StringVar(&userConfig.webhookNamespaceSelector, "webhook-namespace-selector", "", "The webhook will only operate on namespaces with this label, specified in the form key=value. Required if webhook-fail-on-error is true.")
+}
+
 // New creates a new WebHook instance.
 func New(
 	clientset kubernetes.Interface,
 	informerFactory crinformers.SharedInformerFactory,
-	certDir string,
-	webhookServiceNamespace string,
-	webhookServiceName string,
-	webhookPort int,
-	jobNamespace string) (*WebHook, error) {
-	cert := &certBundle{
-		serverCertFile: filepath.Join(certDir, serverCertFile),
-		serverKeyFile:  filepath.Join(certDir, serverKeyFile),
-		caCertFile:     filepath.Join(certDir, caCertFile),
+	jobNamespace string,
+) (*WebHook, error) {
+	cert, err := NewCertProvider(
+		userConfig.serverCert,
+		userConfig.serverCertKey,
+		userConfig.caCert,
+		userConfig.certReloadInterval,
+	)
+	if err != nil {
+		return nil, err
 	}
+
 	path := "/webhook"
 	serviceRef := &v1beta1.ServiceReference{
-		Namespace: webhookServiceNamespace,
-		Name:      webhookServiceName,
+		Namespace: userConfig.webhookServiceNamespace,
+		Name:      userConfig.webhookServiceName,
 		Path:      &path,
 	}
 	hook := &WebHook{
 		clientset:         clientset,
 		lister:            informerFactory.Sparkoperator().V1beta1().SparkApplications().Lister(),
-		cert:              cert,
+		certProvider:      cert,
 		serviceRef:        serviceRef,
 		sparkJobNamespace: jobNamespace,
+		failurePolicy:     arv1beta1.Ignore,
+	}
+	if userConfig.webhookFailOnError {
+		if userConfig.webhookNamespaceSelector == "" {
+			glog.Fatal("webhook-namespace-selector must be set when webhook-fail-on-error is true.")
```

Comment:
Remind me why is this a requirement?

Reply:
Because the operator going down would cause a cluster-wide outage. And if the operator pod were rescheduled, it would required manual intervention (because it would have a circular dependency on itself to start up).

Reply:
Got it. Instead of using `glog.Fatal`, please return an error instead. The caller of `New` will do `glog.Fatal` on the return error anyway.